Skip Missing Param Values

pattisdr opened the pull request:
Closes #2375 

### Code Changes

* [ ] Add a new `SaaSRequest.skip_missing_param_values` attribute with a default value of False 
* [ ] Set the google analytics opt out request to have skip_missing_param_values=true
* [ ] Update SaaSConnector.run_consent_request to skip sending a request if we can't replace all the placeholder values in the request rather than fail, provided skip_missing_param_values=true

### Steps to Confirm

* [ ] Update privacy center test env config.json to have advertising.first_party be executable=true
* [ ] Run `nox -s test_env`
* [ ] Set up the GA consent connector locally
* [ ] Go to the privacy center, delete your ga cookie `_ga="GA1.1.XXXXXXXXX.XXXXXXXXX"`from the browser
* [ ]  In the privacy center, update your consent preferences: Specifically opt out of Email Marketing
* [ ] Go to admin app, approve the privacy request
* [ ] Verify privacy request completes instead of errors
* [ ] Verify in docker container that no request was made to GA, instead you see a log: `Skipping optional consent request on node google_analytics:google_analytics: Unable to replace placeholders in body for the 'opt_out' request of google_analytics`

### Pre-Merge Checklist

* [ ] All CI Pipelines Succeeded
* Documentation:
  * [ ] documentation complete, [PR opened in fidesdocs](https://github.com/ethyca/fidesdocs/pulls)
  * [ ] documentation [issue created in fidesdocs](https://github.com/ethyca/fidesdocs/issues/new/choose)
* [ ] Issue Requirements are Met
* [ ] Relevant Follow-Up Issues Created
* [ ] Update `CHANGELOG.md`

### Description Of Changes

Add a new SaasRequest.skip_missing_param_values to indicate whether we should skip an http request when we can't map all the param values in the request body rather than fail.  

For example, we pull the google analytics client id from the browser for the google analytics consent connector but it isn't always guaranteed to exist.  Rather than having that request fail, we should just skip it.

Functionality added for retrieve_data, mask_data, and run_consent_request, although this is unlikely to be needed for read requests.  (It affects when we can't place param values in the body only)

pattisdr commented on data/saas/config/google_analytics_config.yml:
```diff
@@ -92,6 +92,7 @@ saas_config:
     opt_out:
       method: POST
       path: /analytics/v3/userDeletion/userDeletionRequests:upsert
+      skip_missing_param_values: true
```
This is a patch for the google analytics connector you reviewed previous @adamsachs - wanted to get your thoughts on this approach. Also open to better names!

pattisdr commented on src/fides/api/ops/schemas/saas/saas_config.py:
```diff
@@ -104,6 +104,7 @@ class SaaSRequest(BaseModel):
     grouped_inputs: Optional[List[str]] = []
     ignore_errors: Optional[bool] = False
     rate_limit_config: Optional[RateLimitConfig]
+    skip_missing_param_values: Optional[bool] = False
```
I was toggling between having this param be more specifically related to missing identity data, but thought we might make this more broad to indicate any param values were missing.

pattisdr commented:
I need to retrigger tests too

adamsachs commented:
@pattisdr this looks good to me generally, i like the feature! only question is - since you've put the property on the `SaaSRequest` model generally (which makes sense), should we not also respect it during access/erasure request execution?

pattisdr commented:
@adamsachs thanks so much for taking a look. I've had blinders on trying to get the `consent-propagation` branch ready to merge.  But in the spirit of trying to keep [access, erasure, and consent aligned](https://github.com/ethyca/fides/pull/2302#discussion_r1084538844) we could definitely look into extending this to access/erasure or creating a follow-up ticket at the very least.

pattisdr commented:
@adamsachs I've added a commit to add similar behavior to retrieve_data and mask_data - to have a configurable option to skip the request instead of fail if we can't populate placeholders in the body.

Read requests would be unlikely to hit this error, as this is specifically for **missing placeholders in the body**.  I've added some contrived tests too for read/update requests as we don't have examples yet. 

Let me know what you think!

pattisdr commented:
Realize tests are failing here, but they seem unrelated.  My goal would be to merge this PR into `consent-propgation`, then get `consent-propagation` up-to-date with `main` and I'd re-run tests there.

adamsachs reviewed (approved):
@pattisdr what you added looks great - thanks for addressing that! from my POV, this is good to go 👍
